include/nvexec/stream/reduce.cuh

```diff
@@ -17,221 +17,103 @@
-            return stdexec::get_env(self.op_state_.receiver_);
-          }
+        template <class Range>
+        static void set_value_impl(base::__t&& self, Range&& range) noexcept {
+          cudaStream_t stream = self.op_state_.get_stream();
 
-          __t(Fun fun, operation_state_base_t<ReceiverId>& op_state)
-            : f_((Fun&&) fun)
-            , op_state_(op_state) {
-          }
-        };
-      };
-    }
+          using value_t = result_t<Range>;
+          value_t* d_out = static_cast<value_t*>(self.op_state_.temp_storage_);
 
-    template <class SenderId, class Fun>
-    struct reduce_sender_t {
-      using Sender = stdexec::__t<SenderId>;
+          void* d_temp_storage{};
+          std::size_t temp_storage_size{};
 
-      struct __t : stream_sender_base {
-        using __id = reduce_sender_t;
+          auto first = begin(range);
+          auto last = end(range);
 
-        Sender sndr_;
-        Fun fun_;
+          std::size_t num_items = std::distance(first, last);
 
-        template <class Receiver>
-        using receiver_t =
-          stdexec::__t< reduce_::receiver_t< SenderId, stdexec::__id<Receiver>, Fun>>;
+          cudaError_t status;
 
-        template <class... Range>
-          requires(sizeof...(Range) == 1)
-        using set_value_t = stdexec::completion_signatures<stdexec::set_value_t(
-          std::add_lvalue_reference_t< ::cuda::std::decay_t< ::cuda::std::invoke_result_t<
-            Fun,
-            decltype(*__begin(std::declval<Range>())),
-            decltype(*__begin(std::declval<Range>()))> > >...)>;
-
-        template <class Self, class Env>
-        using completion_signatures = //
-          stdexec::make_completion_signatures<
-            stdexec::__copy_cvref_t<Self, Sender>,
-            Env,
-            stdexec::completion_signatures<stdexec::set_error_t(cudaError_t)>,
-            set_value_t >;
-
-        template <stdexec::__decays_to<__t> Self, stdexec::receiver Receiver>
-          requires stdexec::
-            receiver_of<Receiver, completion_signatures<Self, stdexec::env_of_t<Receiver>>>
-          friend auto
-          tag_invoke(stdexec::connect_t, Self&& self, Receiver&& rcvr) -> stream_op_state_t<
-            stdexec::__copy_cvref_t<Self, Sender>,
-            receiver_t<Receiver>,
-            Receiver> {
-          return stream_op_state<stdexec::__copy_cvref_t<Self, Sender>>(
-            ((Self&&) self).sndr_,
-            (Receiver&&) rcvr,
-            [&](operation_state_base_t<stdexec::__id<Receiver>>& stream_provider)
-              -> receiver_t<Receiver> { return receiver_t<Receiver>(self.fun_, stream_provider); });
-        }
+          do {
+            if (status = STDEXEC_DBG_ERR(cub::DeviceReduce::Reduce(
+                  d_temp_storage,
+                  temp_storage_size,
+                  first,
+                  d_out,
+                  num_items,
+                  self.fun_,
+                  value_t{},
```

Comment:
I guess non-default binary operator doesn't make a lot of sense without an ability to pass initial value. Also, in standard reduce / fold algorithms, there's `reduce(..., init)` and `reduce(..., init, binary_op)` but no `reduce(..., binary_op)`. Do you think you could fix that while we are here?

Reply:
Done